libclamav/regex_list.c

```diff
@@ -162,13 +161,35 @@ cl_error_t regex_list_match(struct regex_matcher *matcher, char *real_url, const
     struct cli_ac_data mdata;
     struct cli_ac_result *res = NULL;
 
-    assert(matcher);
-    assert(real_url);
-    assert(display_url);
+    if (NULL == matcher) {
+        rc = CL_ENULLARG;
+        cli_errmsg("regex_list_match: matcher must be initialized\n");
+        goto done;
+    }
+
+    if (NULL == real_url) {
+        rc = CL_ENULLARG;
+        cli_errmsg("regex_list_match: real_url must be initialized\n");
+        goto done;
+    }
+
+    if (NULL == display_url) {
+        rc = CL_ENULLARG;
+        cli_errmsg("regex_list_match: display_url must be initialized\n");
+        goto done;
+    }
+
     *info = NULL;
-    if (!matcher->list_inited)
-        return CL_SUCCESS;
-    assert(matcher->list_built);
+    if (!matcher->list_inited) {
```

Comment:
I really expected to find that the entire matcher struct was memset to 0, but I can't find anything.  In init_regex_list, matcher is memset to 0, and list_inited is immediately set to 1, so I think you are correct in saying that it is always either 1 or -1.  The variables in the same structure list_built and list_loaded both use 0 and 1, so I agree that changing the errors from -1 to 0 makes sense.

Reply:
In looking at it a little more, I think choices are 0, 1, -1 because it is possible that someone could allocate a regex_matcher struct without calling init_regex_list, and that is why there are checks for being initialized.